ENIGMAsystem/SHELL/Widget_Systems/Cocoa/dialogs.cpp

```diff
@@ -76,34 +82,51 @@ void show_info(string text, int bgcolor, int left, int top, int width, int heigh
 }
 
 int show_message(const string &str) {
-  cocoa_show_message(str.c_str());
-  return 1;
+  if (dialog_caption == "") dialog_caption = cocoa_dialog_caption();
+  return cocoa_show_message(str.c_str(), false, dialog_caption.c_str());
+}
+
+int show_message_cancelable(string str) {
+  if (dialog_caption == "") dialog_caption = cocoa_dialog_caption();
+  return cocoa_show_message(str.c_str(), true, dialog_caption.c_str());
 }
 
 bool show_question(string str) {
-  bool result = (bool)cocoa_show_question(str.c_str());
-  return result;
+  if (dialog_caption == "") dialog_caption = cocoa_dialog_caption();
+  return (bool)cocoa_show_question(str.c_str(), false, dialog_caption.c_str());
+}
+
+int show_question_cancelable(string str) {
+  if (dialog_caption == "") dialog_caption = cocoa_dialog_caption();
+  return cocoa_show_question(str.c_str(), true, dialog_caption.c_str());
+}
+
+int show_attempt(string str) {
+  if (error_caption == "") error_caption = "Error";
+  return cocoa_show_attempt(str.c_str(), error_caption.c_str());
 }
 
 string get_string(string str, string def) {
-  string result = cocoa_input_box(str.c_str(), def.c_str());
-  return result;
+  if (dialog_caption == "") dialog_caption = cocoa_dialog_caption();
```

Comment:
In the future, please prefer `.empty()` to `== ""`.

Reply:
Will do. I can make that change before the merge if you wish. I'll leave it up to you.

Reply:
Is doing what I currently have going to potentially break in the future or something? I'm curious why it's prefered.